File: backend/backend/routers/associacoes.py
from http import HTTPStatus
from typing import Annotated

from fastapi import Depends
from fastapi.routing import APIRouter
from sqlalchemy.ext.asyncio import AsyncSession

from backend.core.engine import EngineApp
from backend.schemas.associacoes import (
    AssociacaoUserContratoCreate,
    AssociacaoUserContratoList,
    AssociacaoUserContratoRead,
    AssociacaoUserEletronicoCreate,
    AssociacaoUserEletronicoList,
    AssociacaoUserEletronicoRead,
)
from backend.security.dependencies import Dependencies, UserContext
from backend.service.associacoes import (
    AssociacaoUserContratoService,
    AssociacaoUserEletronicoService,
)

router_assoc_contrato = APIRouter(
    prefix='/associacoes/contratos',
    tags=['associacoes-contratos'],
)

router_assoc_eletronico = APIRouter(
    prefix='/associacoes/eletronicos',
    tags=['associacoes-eletronicos'],
)

T_AsyncSession = Annotated[AsyncSession, Depends(EngineApp.get_async_session)]
T_UserContext = Annotated[UserContext, Depends(Dependencies.get_user_context)]


# ─── AssociacaoUserContrato


@router_assoc_contrato.get(
    '/',
    status_code=HTTPStatus.OK,
    response_model=AssociacaoUserContratoList,
)
async def get_associacoes_contrato(
    session: T_AsyncSession,
    ctx: T_UserContext,
):
    service = AssociacaoUserContratoService(session)
    associacoes = await service.get(ctx)
    return {'associacoes': associacoes}


@router_assoc_contrato.post(
    '/',
    status_code=HTTPStatus.CREATED,
    response_model=AssociacaoUserContratoRead,
)
async def create_associacao_contrato(
    data: AssociacaoUserContratoCreate,
    session: T_AsyncSession,
    ctx: T_UserContext,
):
    service = AssociacaoUserContratoService(session)
    return await service.create(data, ctx)


@router_assoc_contrato.put(
    '/{user_id}/{centro_custo}',
    status_code=HTTPStatus.OK,
    response_model=AssociacaoUserContratoRead,
)
async def update_associacao_contrato(
    user_id: int,
    centro_custo: str,
    data: AssociacaoUserContratoCreate,
    session: T_AsyncSession,
    ctx: T_UserContext,
):
    service = AssociacaoUserContratoService(session)
    return await service.update(user_id, centro_custo, data, ctx)


@router_assoc_contrato.delete(
    '/{user_id}/{centro_custo}',
    status_code=HTTPStatus.OK,
    response_model=AssociacaoUserContratoRead,
)
async def delete_associacao_contrato(
    user_id: int,
    centro_custo: str,
    session: T_AsyncSession,
    ctx: T_UserContext,
):
    service = AssociacaoUserContratoService(session)
    return await service.delete(user_id, centro_custo, ctx)


# ─── AssociacaoUserEletronico


@router_assoc_eletronico.get(
    '/',
    status_code=HTTPStatus.OK,
    response_model=AssociacaoUserEletronicoList,
)
async def get_associacoes_eletronico(
    session: T_AsyncSession,
    ctx: T_UserContext,
):
    service = AssociacaoUserEletronicoService(session)
    associacoes = await service.get(ctx)
    return {'associacoes': associacoes}


@router_assoc_eletronico.post(
    '/',
    status_code=HTTPStatus.CREATED,
    response_model=AssociacaoUserEletronicoRead,
)
async def create_associacao_eletronico(
    data: AssociacaoUserEletronicoCreate,
    session: T_AsyncSession,
    ctx: T_UserContext,
):
    service = AssociacaoUserEletronicoService(session)
    return await service.create(data, ctx)


@router_assoc_eletronico.delete(
    '/{user_id}/{eletronico_id}',
    status_code=HTTPStatus.OK,
    response_model=AssociacaoUserEletronicoRead,
)
async def delete_associacao_eletronico(
    user_id: int,
    eletronico_id: int,
    session: T_AsyncSession,
    ctx: T_UserContext,
):
    service = AssociacaoUserEletronicoService(session)
    return await service.delete(user_id, eletronico_id, ctx)

File: frontend/frontend/lib/api/associacoes.ts
import { apiFetch } from './client'
import type {
  AssociacaoUserContrato,
  AssociacaoUserEletronico,
  Ocupacao,
} from '@/types/api'

export async function getAssociacoesContrato(): Promise<
  AssociacaoUserContrato[]
> {
  const res = await apiFetch<{ associacoes: AssociacaoUserContrato[] }>(
    '/associacoes/contratos/',
  )
  return res.associacoes
}

export async function createAssociacaoContrato(data: {
  user_id: number
  centro_custo: string
  ocupacao: Ocupacao
}): Promise<AssociacaoUserContrato> {
  return apiFetch<AssociacaoUserContrato>('/associacoes/contratos/', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export async function deleteAssociacaoContrato(
  user_id: number,
  centro_custo: string,
): Promise<AssociacaoUserContrato> {
  return apiFetch<AssociacaoUserContrato>(
    `/associacoes/contratos/${user_id}/${centro_custo}`,
    { method: 'DELETE' },
  )
}

export async function updateAssociacaoContrato(
  user_id: number,
  centro_custo: string,
  ocupacao: 'Gestor' | 'Subgestor' | 'Funcionario',
): Promise<AssociacaoUserContrato> {
  return apiFetch<AssociacaoUserContrato>(
    `/associacoes/contratos/${user_id}/${centro_custo}`,
    {
      method: 'PUT',
      body: JSON.stringify({ user_id, centro_custo, ocupacao }),
    },
  )
}

export async function getAssociacoesEletronico(): Promise<
  AssociacaoUserEletronico[]
> {
  const res = await apiFetch<{ associacoes: AssociacaoUserEletronico[] }>(
    '/associacoes/eletronicos/',
  )
  return res.associacoes
}

export async function createAssociacaoEletronico(data: {
  user_id: number
  eletronico_id: number
}): Promise<AssociacaoUserEletronico> {
  return apiFetch<AssociacaoUserEletronico>('/associacoes/eletronicos/', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export async function deleteAssociacaoEletronico(
  user_id: number,
  eletronico_id: number,
): Promise<AssociacaoUserEletronico> {
  return apiFetch<AssociacaoUserEletronico>(
    `/associacoes/eletronicos/${user_id}/${eletronico_id}`,
    { method: 'DELETE' },
  )
}
